File: server/pptx_parser.py
#!/usr/bin/env python3
"""
Parse a Tony Bianco range review PPTX file.

Layout per slide:
- One heading text box (top-left): contains the LAST name + optional notes
  e.g. "MADDI – LINING – NO TOE PUFF / NO COUNTER – SIZE 11"
  The last name is the first word before any dash/em-dash.
  The heading is identified by cyan (#00FFFF) highlight OR being the topmost text box.

- Multiple style text boxes (one per style column):
  First paragraph = style name line (e.g. "MADDI $199.95" or "MOMA – UNLINED")
  Subsequent paragraphs = SKU lines (colour + leather)

Highlight colours (on <a:highlight> element in run XML):
  #FF00FF  magenta  → specked (sample here)
  #00FFFF  cyan     → specked_no_sample (specked but no sample yet)
  #FFFF00  yellow   → not_specked
  #FF0000  red      → cancelled
  #00FF00  green    → ignore (Anthony's notes)
  none             → carry_over

Output: JSON array to stdout, one object per slide:
  { last, styles: [ { style, skus: [ { colour, leather, status } ] } ] }
"""

import sys
import json
import re

def get_run_highlight(run):
    """Return the hex highlight colour of a run, or None."""
    ns = 'http://schemas.openxmlformats.org/drawingml/2006/main'
    rPr = run._r.find(f'{{{ns}}}rPr')
    if rPr is None:
        return None
    hl = rPr.find(f'{{{ns}}}highlight')
    if hl is None:
        return None
    srgb = hl.find(f'{{{ns}}}srgbClr')
    if srgb is None:
        return None
    val = srgb.get('val', '')
    return '#' + val.upper() if val else None

def get_para_highlight(para):
    """Return the highlight colour of the first highlighted run in a paragraph."""
    for run in para.runs:
        c = get_run_highlight(run)
        if c is not None:
            return c
    return None

def classify_highlight(hex_color):
    """Map hex highlight colour to SKU status."""
    if hex_color is None:
        return 'carry_over'
    h = hex_color.upper()
    if h == '#FF00FF':
        return 'specked'           # magenta = sample here
    if h == '#00FFFF':
        return 'specked_no_sample' # cyan = specked, no sample
    if h == '#FFFF00':
        return 'not_specked'       # yellow = not specked yet
    if h == '#FF0000':
        return 'cancelled'         # red = cancelled
    if h == '#00FF00':
        return 'ignore'            # green = Anthony's notes, ignore
    return 'carry_over'

def extract_last_name(heading_text):
    """Extract last name from heading like 'MADDI – LINING – NO TOE PUFF'."""
    # Split on dash, em-dash, en-dash, dollar
    parts = re.split(r'[\u2013\u2014\-\$]', heading_text)
    name = parts[0].strip()
    # Keep only letters and spaces
    name = re.sub(r'[^A-Za-z\s]', '', name).strip()
    return name.upper() if name else None

def extract_style_name(line):
    """
    Extract style name from a style header line like:
    'MADDI $199.95', 'MOMA – UNLINED', 'MAMZELLE –', 'MAXY $189.95'
    Returns the style name (first word, all caps) or None if not a style line.
    """
    line = line.strip()
    if not line:
        return None
    # Split on dash, em-dash, en-dash, dollar, space+digit
    core = re.split(r'[\u2013\u2014\-\$]|\s+\d', line)[0].strip()
    words = core.split()
    if not words:
        return None
    first = words[0]
    # Must be all uppercase letters, 2-15 chars
    if re.match(r'^[A-Z]{2,15}$', first):
        return first
    return None

def split_colour_leather(text):
    """
    Split 'BLACK NAPPA' → ('BLACK', 'NAPPA')
    Split 'BLACK NAPPA / BLACK PATENT T/C' → ('BLACK', 'NAPPA')
    For two-tone SKUs (e.g. VINTAGE / CHOC UNIT, PATENT / BLACK PATENT TC),
    we use ONLY the primary leather (before the /) so it matches the DB key.
    The secondary material (toe cap, unit, etc.) is a construction detail, not
    a separate leather in the dashboard.
    Returns (colour, leather) or (None, None) if unparseable.
    """
    # First clean up asterisks and trailing notes
    text = re.sub(r'\s*\*.*$', '', text).strip()

    # Split on slash for two-tone — use only the primary (first) part
    slash_parts = [p.strip() for p in text.split('/')]
    primary = slash_parts[0].strip()
    return _split_single(primary)

LEATHER_TYPES = [
    'NAPPA PATENT', 'PATENT TC', 'HI SHINE', 'NAPPA METALLIC',
    'NAPPA', 'SUEDE', 'PATENT', 'CROCO', 'VINTAGE', 'VENICE',
    'NUBUCK', 'KID', 'METALLIC', 'SPECKLE', 'CRINKLE', 'VELVET',
    'SATIN', 'GLITTER', 'MESH', 'FABRIC', 'LEATHER', 'CANVAS',
    'BROCADE', 'WOVEN', 'NYLON', 'SHINE', 'COMO', 'TC',
]

def _split_single(text):
    """Split a single-material string into (colour, leather)."""
    text = text.strip().upper()
    if not text:
        return None, None
    # Try two-word leather types first (longest match)
    words = text.split()
    for i in range(len(words) - 1, -1, -1):
        # Try two-word match
        if i > 0:
            two = words[i-1] + ' ' + words[i]
            if two in LEATHER_TYPES:
                colour = ' '.join(words[:i-1]).strip()
                return (colour if colour else None), two
        # Try one-word match
        if words[i] in LEATHER_TYPES:
            colour = ' '.join(words[:i]).strip()
            return (colour if colour else None), words[i]
    # No leather found
    return None, None

def is_note_line(text):
    """Return True if this line looks like a note/instruction, not a SKU."""
    text = text.strip()
    if not text:
        return True
    # Contains lowercase = likely a note
    if re.search(r'[a-z]', text):
        return True
    # Known note keywords
    note_keywords = ['HEEL', 'HEIGHT', 'LINING', 'TOE PUFF', 'COUNTER',
                     'SIZE', 'CM', 'MM', 'NOTE', 'NB:', 'SEE', 'REFER',
                     'CHECK', 'TBC', 'TBA', 'PENDING', 'CONFIRM',
                     'MICRO STRETCH', 'MICRO STRECH']
    for kw in note_keywords:
        if kw in text.upper():
            return True
    return False

# Words that look like colours but are actually notes/instructions
NOT_COLOURS = {
    'ADD', 'NO', 'TBC', 'TBA', 'SEE', 'NB', 'NOTE', 'CHECK',
    'PENDING', 'CONFIRM', 'REFER', 'ALSO', 'PLUS', 'NEW', 'OLD',
    'YES', 'OR', 'AND', 'THE', 'FOR', 'WITH', 'FROM',
}

def parse_sku_line(text, highlight):
    """Parse a SKU line and return {colour, leather, status} or None."""
    text = text.strip()
    if not text or is_note_line(text):
        return None

    status = classify_highlight(highlight)
    if status == 'ignore':
        return None

    colour, leather = split_colour_leather(text)
    if not colour or not leather:
        return None

    # Reject if colour is a known non-colour instruction word
    if colour.upper() in NOT_COLOURS:
        return None
    # Reject if colour contains question/instruction words (e.g. "SHOULD WE ADD BLACK")
    colour_words = colour.upper().split()
    question_words = {'SHOULD', 'COULD', 'WOULD', 'POSSIBLE', 'MAYBE', 'CONSIDER',
                      'WHAT', 'WHY', 'HOW', 'WHEN', 'WHERE', 'WHICH', 'IF', 'POSSIBLE'}
    if question_words & set(colour_words):
        return None
    # Reject if colour is more than 4 words (likely a note or instruction)
    if len(colour_words) > 4:
        return None

    return {'colour': colour, 'leather': leather, 'status': status}

def parse_slide(slide):
    """Parse a single slide and return {last, styles} or None."""
    ns = 'http://schemas.openxmlformats.org/drawingml/2006/main'

    # Get all text boxes with content, sorted top→bottom, left→right
    boxes = []
    for shape in slide.shapes:
        if shape.has_text_frame and shape.text_frame.text.strip():
            boxes.append(shape)
    if not boxes:
        return None
    boxes.sort(key=lambda s: (s.top, s.left))

    # Find the heading box.
    # The heading is the topmost text box whose FIRST PARAGRAPH looks like a heading
    # (contains a dash/em-dash separator like "DAZIE – YSL HEEL") OR is the topmost box.
    # We deliberately do NOT use cyan highlight to find the heading, because style columns
    # can contain cyan-highlighted SKU lines (specked_no_sample) which would be misidentified
    # as the heading and then skipped during style parsing.
    last_name = None
    heading_box = None
    for box in boxes:
        first_para_text = ''
        for para in box.text_frame.paragraphs:
            t = para.text.strip()
            if t:
                first_para_text = t
                break
        # A heading line contains a dash/em-dash separator (e.g. "DAZIE – YSL HEEL")
        # OR it is the topmost box (fallback)
        has_dash = bool(re.search(r'[\u2013\u2014\-]', first_para_text))
        if has_dash or heading_box is None:
            candidate_last = extract_last_name(first_para_text)
            if candidate_last:
                heading_box = box
                last_name = candidate_last
                if has_dash:
                    break  # found a proper heading with a dash separator
    if not last_name:
        return None
    # Clean last name: take only the first word
    last_name = last_name.split()[0] if last_name.split() else last_name

    # Parse each remaining box as a potential style column
    styles = []
    for box in boxes:
        if box is heading_box:
            continue
        paras = [(p.text.strip(), get_para_highlight(p))
                 for p in box.text_frame.paragraphs
                 if p.text.strip()]
        if not paras:
            continue

        # First paragraph should be the style name
        first_text, first_hl = paras[0]

        # Skip if it looks like a note (heel height etc.)
        if is_note_line(first_text):
            continue

        style_name = extract_style_name(first_text)
        if not style_name:
            continue

        # Skip if style name is a leather/colour word (false positive)
        false_positives = {'HEEL', 'SKIN', 'MILK', 'WEDGE', 'SOLE', 'UPPER',
                           'LINING', 'INSOLE', 'OUTSOLE', 'COUNTER', 'TOE'}
        if style_name in false_positives:
            continue

        skus = []
        for text, highlight in paras[1:]:
            sku = parse_sku_line(text, highlight)
            if sku:
                skus.append(sku)

        styles.append({'style': style_name, 'skus': skus})

    return {'last': last_name, 'styles': styles}

def parse_pptx(filepath):
    from pptx import Presentation
    prs = Presentation(filepath)
    results = []
    for i, slide in enumerate(prs.slides):
        try:
            data = parse_slide(slide)
            if data and data['last']:
                results.append(data)
        except Exception as e:
            results.append({'error': str(e), 'slide': i + 1})
    return results

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print(json.dumps({'error': 'Usage: pptx_parser.py <file.pptx>'}))
        sys.exit(1)
    filepath = sys.argv[1]
    try:
        data = parse_pptx(filepath)
        print(json.dumps(data, ensure_ascii=False))
    except Exception as e:
        print(json.dumps({'error': str(e)}))
        sys.exit(1)

File: server/pptx_parser.ts
/**
 * pptx_parser.ts — Node.js PPTX parser for Tony Bianco range review files.
 *
 * Replicates pptx_parser.py logic using JSZip (to unzip the PPTX) and
 * fast-xml-parser (to read the DrawingML XML). No Python required.
 *
 * Layout per slide:
 *  - One heading text box (top-left): contains the LAST name
 *    Identified by cyan (#00FFFF) highlight OR being the topmost text box.
 *  - Multiple style text boxes (one per style column):
 *    First paragraph = style name line
 *    Subsequent paragraphs = SKU lines (colour + leather)
 *
 * Highlight colours:
 *  #FF00FF  magenta  → specked (sample here)
 *  #00FFFF  cyan     → specked_no_sample
 *  #FFFF00  yellow   → not_specked
 *  #FF0000  red      → cancelled
 *  #00FF00  green    → ignore (Anthony's notes)
 *  none             → carry_over
 */

import JSZip from "jszip";
import { XMLParser } from "fast-xml-parser";

// ─── Constants ────────────────────────────────────────────────────────────────

const LEATHER_TYPES = [
  "NAPPA PATENT", "PATENT TC", "HI SHINE", "NAPPA METALLIC",
  "NAPPA", "SUEDE", "PATENT", "CROCO", "VINTAGE", "VENICE",
  "NUBUCK", "KID", "METALLIC", "SPECKLE", "CRINKLE", "VELVET",
  "SATIN", "GLITTER", "MESH", "FABRIC", "LEATHER", "CANVAS",
  "BROCADE", "WOVEN", "NYLON", "SHINE", "COMO", "TC",
];

const NOT_COLOURS = new Set([
  "ADD", "NO", "TBC", "TBA", "SEE", "NB", "NOTE", "CHECK",
  "PENDING", "CONFIRM", "REFER", "ALSO", "PLUS", "NEW", "OLD",
  "YES", "OR", "AND", "THE", "FOR", "WITH", "FROM",
]);

const QUESTION_WORDS = new Set([
  "SHOULD", "COULD", "WOULD", "POSSIBLE", "MAYBE", "CONSIDER",
  "WHAT", "WHY", "HOW", "WHEN", "WHERE", "WHICH", "IF",
]);

const NOTE_KEYWORDS = [
  "HEEL", "HEIGHT", "LINING", "TOE PUFF", "COUNTER",
  "SIZE", "CM", "MM", "NOTE", "NB:", "SEE", "REFER",
  "CHECK", "TBC", "TBA", "PENDING", "CONFIRM",
  "MICRO STRETCH", "MICRO STRECH",
];

const FALSE_POSITIVE_STYLES = new Set([
  "HEEL", "SKIN", "MILK", "WEDGE", "SOLE", "UPPER",
  "LINING", "INSOLE", "OUTSOLE", "COUNTER", "TOE",
]);

// ─── Types ────────────────────────────────────────────────────────────────────

export type SkuStatus = "specked" | "specked_no_sample" | "not_specked" | "cancelled" | "carry_over" | "ignore";

export interface ParsedSku {
  colour: string;
  leather: string;
  status: SkuStatus;
}

export interface ParsedStyle {
  style: string;
  skus: ParsedSku[];
}

export interface ParsedSlide {
  last: string;
  styles: ParsedStyle[];
  error?: string;
}

// ─── Highlight classification ─────────────────────────────────────────────────

function classifyHighlight(hex: string | null): SkuStatus {
  if (!hex) return "carry_over";
  const h = hex.toUpperCase();
  if (h === "#FF00FF" || h === "FF00FF") return "specked";
  if (h === "#00FFFF" || h === "00FFFF") return "specked_no_sample";
  if (h === "#FFFF00" || h === "FFFF00") return "not_specked";
  if (h === "#FF0000" || h === "FF0000") return "cancelled";
  if (h === "#00FF00" || h === "00FF00") return "ignore";
  return "carry_over";
}

// ─── Text extraction helpers ──────────────────────────────────────────────────

function extractLastName(text: string): string | null {
  // Split on dash, em-dash, en-dash, dollar
  const parts = text.split(/[\u2013\u2014\-\$]/);
  let name = parts[0].trim();
  // Keep only letters and spaces
  name = name.replace(/[^A-Za-z\s]/g, "").trim().toUpperCase();
  // Take only the first word
  const firstWord = name.split(/\s+/)[0];
  return firstWord || null;
}

function extractStyleName(line: string): string | null {
  line = line.trim();
  if (!line) return null;
  // Split on dash, em-dash, en-dash, dollar, space+digit
  const core = line.split(/[\u2013\u2014\-\$]|\s+\d/)[0].trim();
  const words = core.split(/\s+/);
  if (!words.length) return null;
  const first = words[0];
  // Must be all uppercase letters, 2-15 chars
  if (/^[A-Z]{2,15}$/.test(first)) return first;
  return null;
}

function isNoteLine(text: string): boolean {
  text = text.trim();
  if (!text) return true;
  // Contains lowercase = likely a note
  if (/[a-z]/.test(text)) return true;
  // Known note keywords
  const upper = text.toUpperCase();
  for (const kw of NOTE_KEYWORDS) {
    if (upper.includes(kw)) return true;
  }
  return false;
}

function splitColourLeather(text: string): [string | null, string | null] {
  // Clean asterisks and trailing notes
  text = text.replace(/\s*\*.*$/, "").trim();
  // Split on slash for two-tone — use only the primary (first) part
  const primary = text.split("/")[0].trim();
  return splitSingle(primary);
}

function splitSingle(text: string): [string | null, string | null] {
  const words = text.trim().toUpperCase().split(/\s+/);
  if (!words.length) return [null, null];

  // Try two-word leather types first (longest match, scanning right-to-left)
  for (let i = words.length - 1; i >= 0; i--) {
    // Two-word match
    if (i > 0) {
      const two = words[i - 1] + " " + words[i];
      if (LEATHER_TYPES.includes(two)) {
        const colour = words.slice(0, i - 1).join(" ").trim();
        return [colour || null, two];
      }
    }
    // One-word match
    if (LEATHER_TYPES.includes(words[i])) {
      const colour = words.slice(0, i).join(" ").trim();
      return [colour || null, words[i]];
    }
  }
  return [null, null];
}

function parseSkuLine(text: string, highlight: string | null): ParsedSku | null {
  text = text.trim();
  if (!text || isNoteLine(text)) return null;
  const status = classifyHighlight(highlight);
  if (status === "ignore") return null;
  const [colour, leather] = splitColourLeather(text);
  if (!colour || !leather) return null;
  if (NOT_COLOURS.has(colour.toUpperCase())) return null;
  const colourWords = colour.toUpperCase().split(/\s+/);
  if (colourWords.some((w) => QUESTION_WORDS.has(w))) return null;
  if (colourWords.length > 4) return null;
  return { colour, leather, status };
}

// ─── XML parsing helpers ──────────────────────────────────────────────────────

const xmlParser = new XMLParser({
  ignoreAttributes: false,
  attributeNamePrefix: "@_",
  isArray: (name) => ["p:sp", "a:p", "a:r", "a:t", "a:rPr", "a:highlight"].includes(name),
  removeNSPrefix: false,
});

interface RunInfo {
  text: string;
  highlight: string | null;
}

interface ParaInfo {
  text: string;
  highlight: string | null; // first highlighted run
}

interface ShapeInfo {
  top: number;
  left: number;
  paragraphs: ParaInfo[];
}

function parseShapeXml(spXml: any): ShapeInfo | null {
  try {
    // Position
    const spPr = spXml["p:spPr"];
    const xfrm = spPr?.["p:xfrm"] ?? spPr?.["a:xfrm"];
    const off = xfrm?.["a:off"];
    const top = parseInt(off?.["@_y"] ?? "0", 10);
    const left = parseInt(off?.["@_x"] ?? "0", 10);

    // Text frame
    const txBody = spXml["p:txBody"];
    if (!txBody) return null;

    const rawParas = txBody["a:p"];
    if (!rawParas || !Array.isArray(rawParas)) return null;

    const paragraphs: ParaInfo[] = [];

    for (const para of rawParas) {
      const runs: RunInfo[] = [];
      const rawRuns = para["a:r"];
      if (Array.isArray(rawRuns)) {
        for (const run of rawRuns) {
          const tArr = run["a:t"];
          const text = Array.isArray(tArr) ? tArr.map((t: any) => (typeof t === "string" ? t : t["#text"] ?? "")).join("") : (typeof tArr === "string" ? tArr : tArr?.["#text"] ?? "");
          // Highlight from rPr
          let highlight: string | null = null;
          const rPrArr = run["a:rPr"];
          const rPr = Array.isArray(rPrArr) ? rPrArr[0] : rPrArr;
          if (rPr) {
            const hlArr = rPr["a:highlight"];
            const hl = Array.isArray(hlArr) ? hlArr[0] : hlArr;
            if (hl) {
              const srgb = hl["a:srgbClr"];
              if (srgb) {
                const val = srgb["@_val"] ?? (typeof srgb === "string" ? srgb : null);
                if (val) highlight = "#" + val.toUpperCase();
              }
            }
          }
          runs.push({ text: String(text), highlight });
        }
      }

      const fullText = runs.map((r) => r.text).join("").trim();
      const firstHighlight = runs.find((r) => r.highlight)?.highlight ?? null;
      paragraphs.push({ text: fullText, highlight: firstHighlight });
    }

    return { top, left, paragraphs };
  } catch {
    return null;
  }
}

// ─── Slide parser ─────────────────────────────────────────────────────────────

function parseSlide(slideXml: string): ParsedSlide | null {
  const parsed = xmlParser.parse(slideXml);
  const spTree = parsed?.["p:sld"]?.["p:cSld"]?.["p:spTree"];
  if (!spTree) return null;

  const rawShapes = spTree["p:sp"];
  if (!Array.isArray(rawShapes) || rawShapes.length === 0) return null;

  // Parse all shapes
  const shapes: ShapeInfo[] = [];
  for (const sp of rawShapes) {
    const shape = parseShapeXml(sp);
    if (shape && shape.paragraphs.some((p) => p.text)) {
      shapes.push(shape);
    }
  }

  if (shapes.length === 0) return null;

  // Sort top→bottom, left→right
  shapes.sort((a, b) => a.top - b.top || a.left - b.left);

  // Find heading box: prefer cyan-highlighted box, else topmost
  let headingShape: ShapeInfo | null = null;
  let lastName: string | null = null;

  for (const shape of shapes) {
    const hasCyan = shape.paragraphs.some((p) => {
      const h = p.highlight?.toUpperCase();
      return h === "#00FFFF" || h === "00FFFF";
    });
    if (hasCyan || headingShape === null) {
      headingShape = shape;
      const allText = shape.paragraphs.map((p) => p.text).join(" ").trim();
      lastName = extractLastName(allText);
      if (hasCyan) break;
    }
  }

  if (!lastName) return null;

  // Parse each remaining shape as a potential style column
  const styles: ParsedStyle[] = [];

  for (const shape of shapes) {
    if (shape === headingShape) continue;

    const paras = shape.paragraphs.filter((p) => p.text);
    if (paras.length === 0) continue;

    const [firstPara, ...rest] = paras;
    if (isNoteLine(firstPara.text)) continue;

    const styleName = extractStyleName(firstPara.text);
    if (!styleName) continue;
    if (FALSE_POSITIVE_STYLES.has(styleName)) continue;

    const skus: ParsedSku[] = [];
    for (const para of rest) {
      const sku = parseSkuLine(para.text, para.highlight);
      if (sku) skus.push(sku);
    }

    styles.push({ style: styleName, skus });
  }

  return { last: lastName, styles };
}

// ─── Main export ──────────────────────────────────────────────────────────────

export async function parsePptxBuffer(buffer: Buffer): Promise<ParsedSlide[]> {
  const zip = await JSZip.loadAsync(buffer);
  const results: ParsedSlide[] = [];

  // Get slide list from presentation.xml
  const presXml = await zip.file("ppt/presentation.xml")?.async("string");
  if (!presXml) throw new Error("Not a valid PPTX: missing presentation.xml");

  // Find all slide files (ppt/slides/slide1.xml, slide2.xml, ...)
  const slideFiles = Object.keys(zip.files)
    .filter((name) => /^ppt\/slides\/slide\d+\.xml$/.test(name))
    .sort((a, b) => {
      const na = parseInt(a.match(/\d+/)?.[0] ?? "0", 10);
      const nb = parseInt(b.match(/\d+/)?.[0] ?? "0", 10);
      return na - nb;
    });

  for (let i = 0; i < slideFiles.length; i++) {
    const slideFile = slideFiles[i];
    try {
      const xml = await zip.file(slideFile)?.async("string");
      if (!xml) continue;
      const data = parseSlide(xml);
      if (data && data.last) {
        results.push(data);
      }
    } catch (e: any) {
      results.push({ last: "", styles: [], error: `Slide ${i + 1}: ${e.message}` });
    }
  }

  return results;
}
